// @genkit-ai/tools-common/lib/cjs/eval/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractMetricsMetadata = exports.enrichResultsWithScoring = void 0;
const eval_1 = require("../utils/eval");
function enrichResultsWithScoring(scores, evalDataset) {
    const scoreMap = {};
    Object.keys(scores).forEach((evaluator) => {
        const evaluatorResponse = scores[evaluator];
        evaluatorResponse.forEach((scoredSample) => {
            if (!scoredSample.testCaseId) {
                throw new Error('testCaseId expected to be present');
            }
            const score = scoredSample.evaluation;
            if (!scoreMap[scoredSample.testCaseId]) {
                scoreMap[scoredSample.testCaseId] = [];
            }
            scoreMap[scoredSample.testCaseId].push({
                evaluator,
                score: score.score,
                rationale: score.details?.reasoning,
                error: score.error,
                traceId: scoredSample.traceId,
                spanId: scoredSample.spanId,
            });
        });
    });
    return evalDataset.map((evalResult) => {
        return {
            ...evalResult,
            metrics: scoreMap[evalResult.testCaseId] ?? [],
        };
    });
}
exports.enrichResultsWithScoring = enrichResultsWithScoring;
function extractMetricsMetadata(evaluatorActions) {
    const metadata = {};
    for (const action of evaluatorActions) {
        metadata[action.name] = {
            displayName: action.metadata[eval_1.EVALUATOR_METADATA_KEY_DISPLAY_NAME],
            definition: action.metadata[eval_1.EVALUATOR_METADATA_KEY_DEFINITION],
        };
    }
    return metadata;
}
exports.extractMetricsMetadata = extractMetricsMetadata;
