// @genkit-ai/tools-common/lib/esm/eval/parser.js
import { EVALUATOR_METADATA_KEY_DEFINITION, EVALUATOR_METADATA_KEY_DISPLAY_NAME, } from '../utils/eval';
export function enrichResultsWithScoring(scores, evalDataset) {
    const scoreMap = {};
    Object.keys(scores).forEach((evaluator) => {
        const evaluatorResponse = scores[evaluator];
        evaluatorResponse.forEach((scoredSample) => {
            if (!scoredSample.testCaseId) {
                throw new Error('testCaseId expected to be present');
            }
            const score = scoredSample.evaluation;
            if (!scoreMap[scoredSample.testCaseId]) {
                scoreMap[scoredSample.testCaseId] = [];
            }
            scoreMap[scoredSample.testCaseId].push({
                evaluator,
                score: score.score,
                rationale: score.details?.reasoning,
                error: score.error,
                traceId: scoredSample.traceId,
                spanId: scoredSample.spanId,
            });
        });
    });
    return evalDataset.map((evalResult) => {
        return {
            ...evalResult,
            metrics: scoreMap[evalResult.testCaseId] ?? [],
        };
    });
}
export function extractMetricsMetadata(evaluatorActions) {
    const metadata = {};
    for (const action of evaluatorActions) {
        metadata[action.name] = {
            displayName: action.metadata[EVALUATOR_METADATA_KEY_DISPLAY_NAME],
            definition: action.metadata[EVALUATOR_METADATA_KEY_DEFINITION],
        };
    }
    return metadata;
}
